docs/getting_started/module.md

```diff
@@ -0,0 +1,259 @@
+12:53:50.455 kernel  info    stopped background module metadata-server           application: application
+12:53:50.455 kernel  info    stopped stage 1024                                  application: application
+12:53:50.455 kernel  info    stopping stage 0                                    application: application
+12:53:50.455 kernel  info    stopped stage 0                                     application: application
+12:53:50.455 kernel  info    leaving kernel                                      application: application
+```
+
+When running the code, _"foreground-module"_ gets to complete successfully, then as the kernel has no other foreground modules, _"hello-world"_ will have its `ctx` cancelled and thus log `Time to stop`.
+
+## Additional features:
+
+### Default values
+At runtime, every _Module_ object will have to implement the following interfaces:
+
+[embedmd]:# (../../pkg/kernel/module.go /type Module interface/ /\n}/)
+```go
+type Module interface {
+	// Run the module. If the provided context is canceled you have a
+	// few seconds (configurable with kernel.killTimeout) until your module
+	// is killed (via exit(1)). If you return from Run, it is assumed that
+	// your module is done executing and (depending on the type of your module)
+	// this might trigger a kernel shutdown. If you return an error, a kernel
+	// shutdown is also triggered.
+	Run(ctx context.Context) error
+}
+```
+
+This is the backbone of Gosoline module creation. Each _ModuleFactory_ is a function that returns one such _Module_.
+
+[embedmd]:# (../../pkg/kernel/module.go /type TypedModule interface/ /\n}/)
+```go
+type TypedModule interface {
+	IsEssential() bool
+	IsBackground() bool
+}
+```
+
+Internally, a kernel needs to know if a module is a background module, and if it is essential. To configure these properties you need to pass in _Options_ as parameters to `kernel.Add`. The default values, for a module without any specific options, is that this module is a non-essential foreground module. 
+
+[embedmd]:# (../../pkg/kernel/module.go /type StagedModule interface/ /\n}/)
+```go
+type StagedModule interface {
+	GetStage() int
+}
+```
+
+The stage mechanism is explained below.
+
+[embedmd]:# (../../pkg/kernel/module.go /type FullModule interface/ /\n}/)
+```go
+type FullModule interface {
+	Module
+	TypedModule
+	StagedModule
+}
+```
+
+A _FullModule_ needs to implement the other three interfaces.
+
+### Module and stages
+
+When a kernel starts, not all modules are started at the same time. They are grouped into stages, and all modules in a stage start at the same time. When a kernel starts, it begins with starting all modules in stage one, then those in stage two, and so on.
+
+![kernel_lifecycle](module_3.svg)
+
+The stage of a module can be configured with an _Option_, similarly to an essential or background module. If no stage for a module was configured, it will be added to stage 2048.  
```

Comment:
Maybe you should name the stage and not just write its number here. Also, what is the point of stages? That isn't clear after this at all (spoiler: So the metric daemon can collect metrics written by modules when they exit because the application and service stage exit before the essential stage, in which the metric daemon/writer is running.